R/compile_reports.R

```diff
@@ -0,0 +1,159 @@
+#'   factory. Defaults to the current directory.
+#' @param reports Either a regular expression (passed directly to `grep()`) that
+#'   matches to the report paths you would like to compile or an integer/logical
+#'   vector.  If `reports` is an integer or character vector then a call of 
+#'   `compile_reports(factory, reports = idx)` is equivalent to
+#'   `compile_reports(factory, list_reports(factory)[idx])`.
+#' @param params A named list of parameters to be fed to each report that
+#'   matches the regular expression.  If a report has default parameters set in
+#'   it's yaml heading and their are duplicates then preference is given to the
+#'   value supplied here. 
+#' @param quiet A logical indicating if messages from R Markdown compilation
+#'   should be displayed; `TRUE` by default.
+#' @param timestamp timevalue used in the folder structure of the report output.
+#'   If NULL, `format(Sys.time(), "%Y-%m-%d_T%H-%M-%S")` will be used and it's
+#'   evaluation forced immediately.
+#' @param subfolder Name of subfolder to store results.  Not required but helps
+#'   distinguish output if mapping over multiple parameters.
+#' @param ... further arguments passed to `rmarkdown::render()`
+#' 
+#' @importFrom utils write.table
+#' @export 
+compile_reports <- function(factory = ".", reports = NULL,
+                            params = NULL, quiet = TRUE, subfolder = NULL,
+                            timestamp = format(Sys.time(), "%Y-%m-%d_T%H-%M-%S"), 
+                            ...) {
+  
+  # force timestamp to evaluate as soon as function called - needed due to the
+  # `Sys.time` call within the default argument.
+  force(timestamp)
+  
+  # move in to factory
+  tmp <- validate_factory(factory)
+  root <- tmp$root
+  report_sources <- tmp$report_sources
+  outputs <- tmp$outputs
+
+  # get vector of reports to compile
+  report_template_dir <- fs::path(root, report_sources)
+  report_sources <- fs::path(report_template_dir, list_reports(root))
+  if (!is.null(reports)) {
+    if ((is.numeric(reports) && is.wholenumber(reports)) || is.logical(reports)) {
+      report_sources <- report_sources[reports]
+    } else {
+      report_sources <- lapply(reports, grep, report_sources, value = TRUE)
+      report_sources <- unique(unlist(report_sources))
+    }
+  }
+
+  # report output folder
+  report_output_dir <- fs::path(root, outputs)
+  params_to_print <- params
+
+  # loop over all reports
+  for (r in report_sources) {
+
+    # pull yaml from the report
+    yaml <- rmarkdown::yaml_front_matter(r)
+
+    # If params are supplied, these are combined with any default parameters
+    # that may be set in the report header.  Where there are overlaps preference
+    # is given to those set here.  After much trial and error the current way
+    # we facilitate this is to alter the yaml header and save this altered
+    # version to a new file which we then compile.
+    if (!is.null(params)) {
```

Comment:
Given the constraints and multiplte caveats we've seen on this, I think the current solution is actually pretty cool. Who knows, maybe `render` will eventually default to clean environments, in which case we can simplify the code. This said, I really like the trick of changing the yaml. Brutal, but fun!